theano/gof/tests/test_vm.py

```diff
@@ -369,3 +369,93 @@ def check_storage(storage_map):
         assert check_storage(storage_map)[0]
         assert len(set([id(v) for v in
                         storage_map.values()])) < len(storage_map)
+
+
+def test_shape_reallocation():
+        a, b, c, d = [tensor.dvector(n) for n in ['a', 'b', 'c', 'd']]
+        # No Shape Infer
+        z = 3*a + b
+        x = c + 4*d
+        y = z + x
+        # With Shape Infer
+        o = (3*a + b)
+        p = (c + 4*d)
+        w = o + p
+
+        m = theano.compile.get_mode(theano.Mode(linker='vm_nogc'))
+        m = m.excluding('fusion', 'inplace')
+
+        f_1 = theano.function([a, b, c, d], y, name="test_reduce_memory",
+                              mode=m)
+        f_2 = theano.function([a, b, c, d], w, name="test_reduce_memory_i",
+                              mode=m)
+
+        output_1 = f_1([1, 2], [3, 5], [9, 8], [4, 3])
+        output_2 = f_2([1, 2], [3, 5], [9, 8], [4, 3])
+        assert output_1.any()
+        assert output_2.any()
+        storage_map_1 = f_1.fn.storage_map
+        storage_map_2 = f_2.fn.storage_map
+
+        def check_storage(storage_map):
+            from theano.tensor.var import TensorConstant
+            for i in storage_map.keys():
+                if not isinstance(i, TensorConstant):
+                    keys_copy = storage_map.keys()[:]
+                    keys_copy.remove(i)
+                    for o in keys_copy:
+                        if isinstance(
+                            storage_map[i][0] == storage_map[o][0],
+                                numpy.ndarray):
+                            compare = (
+                                storage_map[i][0] == storage_map[o][0]).all()
+                        else:
+                            compare = (storage_map[i][0] == storage_map[o][0])
+                        if storage_map[i][0] is not None and compare:
+                            return [True, storage_map[o][0]]
+            return [False, None]
+
+        assert check_storage(storage_map_1)[0]
+        assert len(set([id(v) for v in
+                        storage_map_1.values()])) < len(storage_map_1)
+        assert check_storage(storage_map_2)[0]
+        assert len(set([id(v) for v in
+                        storage_map_2.values()])) < len(storage_map_2)
+
+
+def test_shape_reallocation_no_shapefeature():
```

Comment:
This test is mostly the same as the one above. The only diff is the mode used and the expected result.

Combine both tests in the same test with a loop over the mode and expected value. This will remove duplicated code.